File: wrappers/javascript/shared/src/crypto/handles.ts
// TODO: do we need these classes?
import { ariesAskar } from '../ariesAskar'

export class ArcHandle {
  public handle: Uint8Array

  public constructor(handle: Uint8Array) {
    this.handle = handle
  }
}

export class StoreHandle {
  public handle: number

  public constructor(handle: number) {
    this.handle = handle
  }

  public async close() {
    await ariesAskar.storeClose({ storeHandle: this })
  }
}

export class ScanHandle {
  public handle: number

  public constructor(handle: number) {
    this.handle = handle
  }

  public free() {
    ariesAskar.scanFree({ scanHandle: this })
  }
}

export class SessionHandle {
  public handle: number

  public constructor(handle: number) {
    this.handle = handle
  }

  public async close(commit: boolean) {
    await ariesAskar.sessionClose({ commit, sessionHandle: this })
  }
}

export class EntryListHandle extends ArcHandle {
  public getCategory(index: number) {
    return ariesAskar.entryListGetCategory({ index, entryListHandle: this })
  }

  public getName(index: number) {
    return ariesAskar.entryListGetName({ index, entryListHandle: this })
  }

  public getValue(index: number) {
    return ariesAskar.entryListGetValue({ index, entryListHandle: this })
  }

  public getTags(index: number) {
    return ariesAskar.entryListGetTags({ index, entryListHandle: this })
  }
}

export class KeyEntryListHandle extends ArcHandle {
  public getAlgorithm(index: number) {
    return ariesAskar.keyEntryListGetAlgorithm({ index, keyEntryListHandle: this })
  }

  public getName(index: number) {
    return ariesAskar.keyEntryListGetName({ index, keyEntryListHandle: this })
  }

  public getTags(index: number) {
    return ariesAskar.keyEntryListGetTags({ index, keyEntryListHandle: this })
  }

  public getMetadata(index: number) {
    return ariesAskar.keyEntryListGetMetadata({ index, keyEntryListHandle: this })
  }

  public loadKey(index: number) {
    return ariesAskar.keyEntryListLoadLocal({ index, keyEntryListHandle: this })
  }
}

export class LocalKeyHandle extends ArcHandle {}

File: wrappers/kotlin/src/commonMain/kotlin/crypto/handles.kt
@file:OptIn(UnsafeNumber::class)

package askar.crypto

import aries_askar.*
import aries_askar.EntryListHandle
import aries_askar.KeyEntryListHandle
import aries_askar.LocalKeyHandle
import aries_askar.ScanHandle
import aries_askar.SessionHandle
import aries_askar.StoreHandle
import askar.Askar
import kotlinx.cinterop.*
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking


class StoreHandle(val handle: StoreHandle) {

    suspend fun close() {
        Askar.store.storeClose(this.handle)
    }

    companion object {
        fun fromHandle(handle: StoreHandle?): askar.crypto.StoreHandle? {
            if (handle != null)
                return StoreHandle(handle)
            return null
        }
    }
}

class ScanHandle(val handle: ScanHandle) {

    fun free() {
        Askar.scan.free(this.handle)
    }

    companion object {
        fun fromHandle(handle: ScanHandle?): askar.crypto.ScanHandle? {
            if (handle != null)
                return ScanHandle(handle)
            return null
        }
    }
}

class SessionHandle(val handle: SessionHandle) {

    suspend fun close(commit: Boolean): Long {
        val h = this.handle
        val code: Long
        runBlocking {
            val c = async {Askar.session.sessionClose(h, commit)}
            code = c.await()
        }
        return code
    }


    companion object {
        fun fromHandle(handle: SessionHandle?): askar.crypto.SessionHandle? {
            if (handle != null)
                return SessionHandle(handle)
            return null
        }
    }
}

class EntryListHandle(val handle: EntryListHandle) {

    fun getCategory(index: Int): String {
        return Askar.entryList.getCategory(index, this.handle)
    }

    fun getName(index: Int): String {
        return Askar.entryList.getName(index, this.handle)
    }

    fun getValue(index: Int): String {
        return Askar.entryList.getValue(index, this.handle)
    }

    fun getTags(index: Int): String? {
        return Askar.entryList.getTags(index, handle)
    }

    fun free() {
        Askar.entryList.free(this.handle)
    }

    companion object {
        fun fromHandle(handle: EntryListHandle?): askar.crypto.EntryListHandle? {
            if(handle?._0 == null)
                return null
            return EntryListHandle(handle)
        }
    }
}

class KeyEntryListHandle(val handle: KeyEntryListHandle) {

    fun getAlgorithm(index: Int): String {
        return Askar.keyEntryList.getAlgorithm(index, this.handle)
    }

    fun getName(index: Int): String {
        return Askar.keyEntryList.getName(index, this.handle)
    }

    fun getTags(index: Int): String? {
        return Askar.keyEntryList.getTags(index, this.handle)
    }

    fun getMetadata(index: Int): String? {
        return Askar.keyEntryList.getMetadata(index, this.handle)
    }

    fun loadKey(index: Int): askar.crypto.LocalKeyHandleKot {
        return LocalKeyHandleKot(Askar.keyEntryList.loadLocal(index, this.handle))
    }

    fun free() {
        Askar.keyEntryList.free(this.handle)
    }

    companion object {
        fun fromHandle(handle: KeyEntryListHandle?): askar.crypto.KeyEntryListHandle? {
            if(handle != null)
                return KeyEntryListHandle(handle)
            return null
        }
    }
}
class LocalKeyHandleKot(val handle: LocalKeyHandle)  {

    fun free() {
        Askar.key.keyFree(this.handle)
        nativeHeap.free(handle)
    }

    companion object {
        fun fromHandle(handle: LocalKeyHandle?): askar.crypto.LocalKeyHandleKot? {
            if (handle != null)
                return LocalKeyHandleKot(handle)
            return null
        }
    }

//    object LocalKeyHandleAsStringSerializer : KSerializer<LocalKeyHandle> {
//        override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("handle", PrimitiveKind.STRING)
//        override fun serialize(encoder: Encoder, value: LocalKeyHandle) {
//            val count = value._0!!.reinterpret<FfiStrVar>().pointed.value!!.toKString().length
//            println(count)
//            val temp = value._0!!.pointed
//            val s = value._0!!.reinterpret<ByteVar>()
//            println()
//            encoder.encodeString("test")
//        }
//
//        override fun deserialize(decoder: Decoder): LocalKeyHandle {
//            val s = decoder.decodeString()
//            val h = cValue<LocalKeyHandle> {
//                //TODO: s needs to be converted back to a pointer
//               _0 = null
//            }
//            h.useContents {
//                return this
//            }
//        }
//    }


}





